### PR TITLE
Files additionalscripts to real event

### PR DESCRIPTION
**Pull request by rullzer**
* New additionalscripts event
* Legacy handler
* Moved comments as a POC

This allows to write nice tiny classes that just do what they need instead of having code and connetion mixed.

### CODE REVIEW

File: apps/files/lib/Listener/LegacyLoadAdditionalScripts.php [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+
+namespace OCA\Files\Listener;
+
+use OCA\Files\Event\LoadAdditionalScriptsEvent;
+use OCP\EventDispatcher\Event;
+use OCP\EventDispatcher\IEventListener;
+use Symfony\Component\EventDispatcher\EventDispatcherInterface;
+use Symfony\Component\EventDispatcher\GenericEvent;
+
+class LegacyLoadAdditionalScripts implements IEventListener {
```
Comment (ChristophWurst) [inline apps/files/lib/Listener/LegacyLoadAdditionalScripts.php]:
```suggestion
class LegacyLoadAdditionalScriptsAdapter implements IEventListener {
```

File: apps/files/lib/Listener/LegacyLoadAdditionalScripts.php [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+
+	/** @var EventDispatcherInterface */
+	private $legacyDispatcher;
+
+	public function __construct(EventDispatcherInterface $legacyDispatcher) {
```
Comment (ChristophWurst) [inline apps/files/lib/Listener/LegacyLoadAdditionalScripts.php]: you can also use the new one. it extends the original one, so there is just one instance

**Comment by ChristophWurst** [comment]
oh an conflicts :)

**Comment by rullzer** [comment]
updated :)

File: apps/comments/lib/AppInfo/Application.php [hunk 3]
```diff
@@ -49,21 +52,18 @@ public function register() {
...
 
-	protected function registerSidebarScripts(EventDispatcherInterface $dispatcher) {
-		$dispatcher->addListener(
-			'OCA\Files::loadAdditionalScripts',
-			function() {
-				Util::addScript('comments', 'comments');
-			}
-		);
+	protected function registerSidebarScripts(IEventDispatcher $dispatcher) {
+		$dispatcher->addServiceListener(LoadAdditionalScriptsEvent::class, LoadAdditionalScripts::class);
```
Comment (ChristophWurst) [inline apps/comments/lib/AppInfo/Application.php]: for one-liners you could also use the `addListener` and pass a closure. but fine by me to have it in a class :)

File: apps/files/lib/Listener/LegacyLoadAdditionalScriptsAdapter.php [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+use OCP\EventDispatcher\Event;
+use OCP\EventDispatcher\IEventDispatcher;
+use OCP\EventDispatcher\IEventListener;
+
+class LegacyLoadAdditionalScriptsAdapter implements IEventListener {
+
+	/** @var IEventDispatcher */
+	private $dispatcher;
+
+	public function __construct(IEventDispatcher $legacyDispatcher) {
```
Comment (ChristophWurst) [inline apps/files/lib/Listener/LegacyLoadAdditionalScriptsAdapter.php]:
```suggestion
	public function __construct(IEventDispatcher $dispatcher) {
```

File: apps/files/lib/Listener/LegacyLoadAdditionalScriptsAdapter.php [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+		$this->dispatcher = $legacyDispatcher;
```
Comment (ChristophWurst) [inline apps/files/lib/Listener/LegacyLoadAdditionalScriptsAdapter.php]:
```suggestion
		$this->dispatcher = $dispatcher;
```

**Review by ChristophWurst — CHANGES_REQUESTED** [review]

**Comment by rullzer** [comment]
Lets do this for 18 (early) as we need other events as well and more functionality.

**Comment by nickvergessen** [comment]
I would also prefer to merge the 2 event dispatchers if easily possible (one wrapping the other or somthing)

**Comment by ChristophWurst** [comment]
> I would also prefer to merge the 2 event dispatchers if easily possible (one wrapping the other or somthing)

[That is already the case](https://github.com/nextcloud/server/blob/master/lib/private/EventDispatcher/EventDispatcher.php) using composition and delegation. There is only one instance of a dispatcher, but two interfaces for it.

**Comment by kesselb** [comment]
image

**Review by ChristophWurst — APPROVED** [review]

**Comment by rullzer** [comment]
@nickvergessen please have another look. cleaner now

**Review by icewind1991 — APPROVED** [review]

**Comment by rullzer** [comment]
Recent new eventdispatcher edits broke this. I'll have to take another look.

**Review by ChristophWurst — APPROVED** [review]

**Comment by nickvergessen** [comment]
So, ready to merge?

**Pull request merged**